rmc/src/service.rs

```diff
@@ -0,0 +1,420 @@
+        >,
+        network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+        scheduler: impl TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>> + 'static,
+    ) -> Self {
+        Service {
+            network_rx,
+            network_tx,
+            scheduler: Box::new(scheduler),
+        }
+    }
+
+    fn handle_message(
+        &mut self,
+        message: RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+        handler: &mut Handler<H, MK>,
+    ) {
+        match message {
+            RmcIncomingMessage::StartRmc(hash) => {
+                debug!(target: "AlephBFT-rmc", "starting rmc for {:?}", hash);
+                let unchecked = handler.on_start_rmc(hash);
+                self.scheduler.add_task(RmcHash::SignedHash(unchecked));
+            }
+            RmcIncomingMessage::RmcHash(RmcHash::MultisignedHash(unchecked)) => {
+                match handler.on_multisigned_hash(&unchecked) {
+                    Ok(Some(multisigned)) => {
+                        self.scheduler.add_task(RmcHash::MultisignedHash(unchecked));
+                        if self
+                            .network_tx
+                            .unbounded_send(RmcOutgoingMessage::NewMultisigned(multisigned))
+                            .is_err()
+                        {
+                            error!(target: "AlephBFT-rmc", "network closed early.");
+                        }
+                    }
+                    Ok(None) => {}
+                    Err(error) => {
+                        warn!(target: "AlephBFT-rmc", "{}", error);
+                    }
+                }
+            }
+            RmcIncomingMessage::RmcHash(RmcHash::SignedHash(unchecked)) => {
+                match handler.on_signed_hash(&unchecked) {
+                    Ok(Some(multisigned)) => {
+                        self.scheduler.add_task(RmcHash::MultisignedHash(
+                            multisigned.clone().into_unchecked(),
+                        ));
+                        if self
+                            .network_tx
+                            .unbounded_send(RmcOutgoingMessage::NewMultisigned(multisigned))
+                            .is_err()
+                        {
+                            error!(target: "AlephBFT-rmc", "network closed early.");
+                        }
+                    }
+                    Ok(None) => {}
+                    Err(error) => {
+                        warn!(target: "AlephBFT-rmc", "{}", error);
+                    }
+                }
+            }
+        }
+    }
+
+    fn send_hash(&self, hash: RmcHash<H, MK::Signature, MK::PartialMultisignature>) {
+        self.network_tx
+            .unbounded_send(RmcOutgoingMessage::RmcHash(hash))
+            .expect("Sending message should succeed");
+    }
+
+    /// Run the rmc service.
+    pub async fn run(mut self, mut handler: Handler<H, MK>, mut terminator: Terminator) {
```

Comment:
Handler can't be just part of the `Service` type?

Reply:
yeah, I think it can. the same should apply for `alerter` so imma solve it too